src/main/java/io/github/syst3ms/skriptparser/expressions/ExprAmount.java

```diff
@@ -1,46 +1,90 @@
 package io.github.syst3ms.skriptparser.expressions;
 
 import io.github.syst3ms.skriptparser.Main;
-import io.github.syst3ms.skriptparser.lang.TriggerContext;
 import io.github.syst3ms.skriptparser.lang.Expression;
+import io.github.syst3ms.skriptparser.lang.Literal;
+import io.github.syst3ms.skriptparser.lang.TriggerContext;
+import io.github.syst3ms.skriptparser.lang.Variable;
+import io.github.syst3ms.skriptparser.lang.base.PropertyExpression;
+import io.github.syst3ms.skriptparser.log.ErrorType;
 import io.github.syst3ms.skriptparser.parsing.ParseContext;
+import org.jetbrains.annotations.NotNull;
 import org.jetbrains.annotations.Nullable;
 
+import java.util.Map;
+
 /**
- * Amount of a list of values.
+ * The amount of elements in a given list.
+ * Note that when getting the size of a list variable, it will only return the size of the first layer of elements.
+ * If you want to get the whole list size, including nested layers in a variable, use the recursive size instead.
  *
  * @name Amount
- * @pattern (amount|number|size) of %objects%
+ * @type EXPRESSION
+ * @pattern [the] [recursive] (amount|number|size) of %objects%
+ * @pattern %objects%'[s] [recursive] (amount|number|size)
  * @since ALPHA
- * @author Olyno
+ * @author Olyno, Mwexim
  */
-public class ExprAmount implements Expression<Number> {
-
-    private Expression<Object> valuesList;
-
-    static {
-        Main.getMainRegistration().addExpression(
-            ExprAmount.class,
-            Number.class,
-            true,
-            "(amount|number|size) of %objects%"
-        );
-    }
-
-    @SuppressWarnings("unchecked")
-    @Override
-    public boolean init(Expression<?>[] expressions, int matchedPattern, ParseContext parseContext) {
-        valuesList = (Expression<Object>) expressions[0];
-        return true;
-    }
-
-    @Override
-    public Number[] getValues(TriggerContext ctx) {
-        return new Number[]{valuesList.getValues(ctx).length};
-    }
-
-    @Override
-    public String toString(@Nullable TriggerContext ctx, boolean debug) {
-        return "amount of " + valuesList.toString(ctx, debug);
-    }
+public class ExprAmount extends PropertyExpression<Number, Object> {
+
+	static {
+		Main.getMainRegistration().addPropertyExpression(
+				ExprAmount.class,
+				Number.class,
+				true,
+				"objects",
```

Comment:
To restrict the possible expressions to non-literals only, use `"~objects"` instead. That removes the need for lines 45 through 51.